File: core/src/main/scala/com/graphbrain/eco/nodes/MaxDepthFun.scala
package com.graphbrain.eco.nodes

import com.graphbrain.eco.NodeType.NodeType
import com.graphbrain.eco.{Contexts, NodeType}

class MaxDepthFun(params: Array[ProgNode], lastTokenPos: Int= -1) extends FunNode(params, lastTokenPos) {
  override val label = "max-depth"

  override def ntype: NodeType = NodeType.Boolean

  override def booleanValue(ctxts: Contexts) = {
    params(0).numberValue(ctxts)

    for (c <- ctxts.ctxts) {
      val r = ctxts.depth <= c.getRetNumber(params(0))
      c.setRetBoolean(this, r)
    }
  }

  override protected def typeError() = error("parameters must be boolean")
}

File: core/src/main/java/com/graphbrain/eco/nodes/MaxDepthFun.java
package com.graphbrain.eco.nodes;

import com.graphbrain.eco.Context;
import com.graphbrain.eco.Contexts;
import com.graphbrain.eco.NodeType;

public class MaxDepthFun extends FunNode {

    public MaxDepthFun(ProgNode[] params, int lastTokenPos) {
        super(params, lastTokenPos);
    }

    public MaxDepthFun(ProgNode[] params) {
        this(params, -1);
    }

    @Override
    public String label(){return "max-depth";}

    @Override
    public NodeType ntype(Context ctxt){return NodeType.Boolean;}

    @Override
    public void eval(Contexts ctxts) {
        params[0].eval(ctxts);

        for (Context c : ctxts.getCtxts()) {
            boolean r = ctxts.getDepth() <= c.getRetNumber(params[0]);
            c.setRetBoolean(this, r);
        }
    }
}